source/common/filter/ext_authz.cc

```diff
@@ -0,0 +1,101 @@
+  const std::string final_prefix = fmt::format("ext_authz.{}.", name);
+  return {ALL_TCP_EXT_AUTHZ_STATS(POOL_COUNTER_PREFIX(scope, final_prefix),
+                                  POOL_GAUGE_PREFIX(scope, final_prefix))};
+}
+
+void Instance::callCheck() {
+  CheckRequestUtils::createTcpCheck(filter_callbacks_, checkRequest_);
+
+  status_ = Status::Calling;
+  filter_callbacks_->connection().readDisable(true);
+  config_->stats().active_.inc();
+  config_->stats().total_.inc();
+
+  calling_check_ = true;
+  client_->check(*this, checkRequest_, Tracing::NullSpan::instance());
+  calling_check_ = false;
+}
+
+Network::FilterStatus Instance::onData(Buffer::Instance&) {
+  if (status_ == Status::NotStarted) {
+    // If there was no ssl handshake there will be no event Network::ConnectionEvent::Connected
+    // and onData() will be the first event into the filter.
+    // Therefore, if the authorization service hasn not been invoked then we must do so now.
+    callCheck();
+  }
+  return status_ == Status::Calling ? Network::FilterStatus::StopIteration
+                                    : Network::FilterStatus::Continue;
+}
+
+Network::FilterStatus Instance::onNewConnection() {
+  // Wait till the next event occurs.
+  return Network::FilterStatus::Continue;
+}
+
+void Instance::onEvent(Network::ConnectionEvent event) {
+  if (event == Network::ConnectionEvent::RemoteClose ||
+      event == Network::ConnectionEvent::LocalClose) {
+    if (status_ == Status::Calling) {
+      // Make sure that any pending request in the client is cancelled. This will be NOP if the
+      // request already completed.
+      client_->cancel();
+      config_->stats().active_.dec();
+    }
+  } else {
+    // SSL connection is post TCP newConnection. Therefore the ext_authz check in onEvent.
+    // if the ssl handshake was successful then it will invoke the
+    // Network::ConnectionEvent::Connected.
+    // Thus invoke the authorization service check if not done yet.
+    if (status_ == Status::NotStarted) {
+      callCheck();
+    }
+  }
+}
+
+void Instance::onComplete(CheckStatus status) {
+  status_ = Status::Complete;
+  filter_callbacks_->connection().readDisable(false);
+  config_->stats().active_.dec();
+
+  switch (status) {
+  case CheckStatus::OK:
+    config_->stats().ok_.inc();
+    break;
+  case CheckStatus::Error:
+    config_->stats().error_.inc();
+    break;
+  case CheckStatus::Denied:
+    config_->stats().denied_.inc();
+    break;
+  }
+
+  // Fail open only if configured to do so and if the check status was a error.
+  if (status == CheckStatus::Denied || (status == CheckStatus::Error && !config_->failOpen())) {
+    config_->stats().cx_closed_.inc();
+    filter_callbacks_->connection().close(Network::ConnectionCloseType::NoFlush);
+  } else {
+    // We can get completion inline, so only call continue if that isn't happening.
+    if (!calling_check_) {
+      filter_callbacks_->continueReading();
```

Comment:
I wonder if `readDisable(false)` above is sufficient, and you don't need to do this; it's docs say it will restart the read pipeline, and redispatch pending data. Does this work for this filter? That could eliminate the need to maintain `calling_check_`.

Reply:
Moving `readDisable(false)` here -which is what u were asking about at the top of this function-makes sense.
From what I understood-which may be totally wrong-is that  `readDisable(false)` was to get the socket buffer pipeline moving and the `filter_callbacks_->continueReading()` is to get the filter chain iteration moving forward. The filter iteration may have been stopped iff the `callCheck()`and the `onComplete()` callback happened asynchronously.
But if they both are invoked in the same call stack then the filter chain was never stopped, hence it doesn't need to be kick started again by invoking `filter_callbacks_->continueReading()`.

Reply:
I think that `filter_callback->connection().readDisable(false|true)` will no longer be necessary since now we will invoke the external authz service from `onData()` which clearly returns either a  `StopIteration` or `Continue` and NOT from `onEvent()`.